### PR TITLE
ci: Support secure binary signing on protected pipelines

### PR DESCRIPTION
**Pull request by scottwittenburg**
This PR supports the creation of securely signed binaries built from spack `develop` as well as release branches and tags.  Specifically:

- remove internal pr mirror url generation logic in favor of buildcache destination on command line
    - with a single mirror url specified in the spack.yaml,  this makes it clearer where binaries from various pipelines are pushed
- designate some tags as reserved: ['public', 'protected', 'notary']
    - these tags are stripped from all jobs by default and provisioned internally based on pipeline type
- update gitlab ci yaml to include pipelines on more protected branches than just develop (so include releases and tags)
    - binaries from all protected pipelines are pushed into mirrors including the branch name so releases, tags, and develop binaries are kept separate
- update rebuild jobs running on protected pipelines to run on special runners provisioned with an intermediate signing key
    - protected rebuild jobs no longer use "SPACK_SIGNING_KEY" env var to obtain signing key (in fact, final signing key is nowhere available to rebuild jobs)
    - these itermediate signatures are verified at the end of each pipeline by a new signing job to ensure binaries were produced by a protected pipeline
- optionallly schedule a signing/notary job at the end of the pipeline to sign all packges in the mirror
    - add signing-job-attributes to gitlab-ci section of spack environment to allow configuration
    - signing job runs on special runner (separate from protected rebuild runners) provisioned with public intermediate key and secret signing key

### CODE REVIEW

**Comment by scottwittenburg** [comment]
@tgamblin Here's the other PR we need for public signed binaries.

@kotfic @zackgalbreath @opadron

**Review by alalazo — CHANGES_REQUESTED** [review]
```
lib/spack/spack/ci.py:1220: [E225] missing whitespace around operator
lib/spack/spack/test/cmd/ci.py:1748: [E111] indentation is not a multiple of 4
flake8 found errors
```

File: lib/spack/spack/ci.py
Comment (alalazo) [inline lib/spack/spack/ci.py:611]: (Super minor, just OCD) Never noticed but we have a typo below "dynamic :chile: pipeline"

File: lib/spack/spack/ci.py
Comment (alalazo) [inline lib/spack/spack/ci.py:698]: This function is ~700 lines. Is it feasible, in the short time frame we have before v0.18, to break it in smaller logical chunks?

File: lib/spack/spack/ci.py
Comment (alalazo) [inline lib/spack/spack/ci.py]: I think we perform this filtering enough times that we could benefit from extracting it into a function

File: lib/spack/spack/test/cmd/ci.py
Comment (alalazo) [inline lib/spack/spack/test/cmd/ci.py]: Can this block be un-indented one level?

File: lib/spack/spack/test/cmd/ci.py
Comment (alalazo) [inline lib/spack/spack/test/cmd/ci.py]: Very minor OCD, but we don't need parenthesis:
```suggestion
                assert 'sign-pkgs' in pipeline_doc
```

**Review by alalazo — CHANGES_REQUESTED** [review]
I just left a few minor comments on the code. It would be good if somebody could double check the pipelines. I'm not super familiar with them, so I might miss issues more easily.

If breaking the function is hard / requires time let's make an issue about that to be solved after the release. I think a 700 lines function is a tech debt that need to be solved (to be clear, this was not introduced by this PR).

File: lib/spack/spack/ci.py
Comment (scottwittenburg) [reply lib/spack/spack/ci.py:698]: I think anything I try to do this week is going to be rushed and may just have to be revisited again shortly anyway.  If you're ok waiting until after the pending deadlines, there's a lot of refactoring I'd like to tackle in here.

Comment (tgamblin) [reply lib/spack/spack/ci.py:698]: I'm ok with holding off on the refactor here, but we really do need to clean this stuff up soon after 0.18, as there's getting to be a lot of tech debt.

I think the interesting question here is how best to separate the stuff we need for production pipelines and for pipelines created by end users.  It seems like the specifics of prod are creeping in here, and I think it makes it harder to generify the code.

**Review by tgamblin — DISMISSED** [review]
This LGTM, modulo @alalazo's request to refactor, which I think we need to hold off on given the time constraints.  I don't see anything absolutely horrible, though it might be good to get a once-over from @opadron, @zackgalbreath, or @kotfic as you all know the pipelines a lot better than I do at this point.

**Review by alalazo — DISMISSED** [review]
Every point in the previous review was addressed but refactor, which should be done after v0.18

**Review by zackgalbreath — DISMISSED** [review]

**Comment by scottwittenburg** [comment]
This should get merged *after* #30750, and should be coordinated with changes to the sync script and some actions in gitlab.

**Comment by scottwittenburg** [comment]
Looking into the failure of `lib/spack/spack/test/cmd/checksum.py::test_checksum` which for sure didn't fail on my machine, but it seems failed in CI.

I'm just trying a re-run of the failed jobs due to how the stack trace went into python code attempting to read from a url and died with some kind of ssl error deep inside.

Is that test known to fail spuriously now and then?

**Comment by scottwittenburg** [comment]
@eugeneswalker Do you want to have a look for how the `e4s-mac` stack has been handled here?

**Comment by eugeneswalker** [comment]
> @eugeneswalker Do you want to have a look for how the `e4s-mac` stack has been handled here?

LGTM! Thank you!

**Review by zackgalbreath — APPROVED** [review]

**Pull request merged**
